### PR TITLE
chore: common unified admin page

### LINKED ISSUE #919: Add video permissions management for organiser teams and fix access control in unified system
With the **video component** now integrated into the unified eventyay system, organisers and their teams need appropriate access control to manage video features.
Currently, the permission system does not behave as expected, and organisers cannot reliably manage video areas for their events.

### **Expected Behavior:**

1. **System admins** have access to all video sections across all events.
2. **Event organiser admin teams (event creators)** have full access to the video area of their own events.
3. **Attendees** can view video event areas (as viewers only).
4. **Organisers (admin teams/event creators)** should be able to control video permissions when creating or editing teams.

   * (Compare to team permission interface at:
     `https://wikimedia.eventyay.com/tickets/control/organizer/{organisername}/team/{teamnumber}/edit`)

### **Proposed Solution:**

* Add a new section titled **“Video permissions”** to the **team permissions page**.
* Include the following permission options under that heading:

  * [ ] Can create Stages
  * [ ] Can create Channels
  * [ ] Can send Direct Messages
  * [ ] Can create Announcements
  * [ ] Can view Users
  * [ ] Can message, ban, and silence Users
  * [ ] Can create and edit Rooms
  * [ ] Can create and edit Kiosks
  * [ ] Can edit Event Configuration

### **Acceptance Criteria:**

* [ ] Admins have unrestricted video access globally.
* [ ] Organiser teams can fully manage video features for their own events.
* [ ] Attendees retain view-only permissions.
* [ ] “Video permissions” section visible and functional in team edit pages.
* [ ] Video-related features respect the updated permission model.

---

### **Notes:**

Ensure that all video-related routes and APIs (stages, channels, chat, etc.) reference the new permission checks.
The system should handle **new teams** created under an organiser.

Image

### LINKED ISSUE #921: Unify team creation and management pages under a single organiser permissions interface
The current team management structure under the organiser section is fragmented across multiple subpages and components, making it harder to manage teams and their permissions.
To simplify navigation and create a unified permissions management experience, all team creation and permission management pages should be consolidated into a single view under the organiser section.

#### **Step 1: Move Teams Section**

* Move the current team pages from
  `https://wikimedia.eventyay.com/tickets/control/organizer/{organisername}/teams`
  and its subpages
  to
  `https://wikimedia.eventyay.com/tickets/common/organizer/{organisername}`
* **Remove the `/update` slug** from the URL to make it cleaner.

#### **Step 2: List All Teams**

* On the new unified page, list all teams the same way they appear on
  `https://wikimedia.eventyay.com/tickets/control/organizer/{organisername}/teams/{teamnumber}`

#### **Step 3: Integrate Permissions Management**

* On the same page, include a list of all team permissions (as shown on the edit page):
  `https://wikimedia.eventyay.com/tickets/control/organizer/{organisername}/teams/{teamnumber}/edit`
* This effectively **merges the team overview and team permissions edit page**.
* After implementation, remove the separate `/edit` page.

#### **Step 4: Ensure Integration with Related Permissions**

* Confirm that the unified team management interface fully integrates with:

  * **Talk permissions** → #920
  * **Video permissions** → #919

### **Expected Outcome:**

* All team creation, management, and permission editing functions are accessible in one place.
* The URL structure is simplified and consistent under `/tickets/common/organizer/{organisername}`.
* `/update` and `/edit` slugs are removed for cleaner navigation.
* Permissions for talk and video features are seamlessly integrated into the unified page.

### **Acceptance Criteria:**

* [ ] Teams and permissions appear on a single organiser management page.
* [ ] Old `/teams` and `/edit` routes are deprecated.
* [ ] Talk and video permissions are visible and functional on the unified interface.
* [ ] Navigation and layout are consistent with other organiser management pages.

### PR DESCRIPTION
**Pull request by Saksham-Sirohi**
Fixes #921
Fixes #919

## Summary by Sourcery

Unify organizer administration by moving all team and video permission management into the organizer settings page, deprecating legacy team-specific views/URLs, and tightening video access control across backend and frontend.

New Features:
- Introduce a tabbed organizer management page that combines general organizer settings with team creation and inline management of members, invites, and API tokens.
- Add granular video-related team permissions that map to specific Eventyay Video roles and traits for event worlds.

Bug Fixes:
- Restrict Eventyay Video admin-level traits in JWTs to staff users only and base other video traits on team permissions.
- Ensure direct messaging, room creation, kiosk management, and moderation actions in the video frontend are only available when the corresponding permissions are present.

Enhancements:
- Refactor organizer update handling to process team CRUD, member, invite, and token operations in a scoped, transactional way with audit logging of changes.
- Centralize and reuse team permission fieldsets between organizer and team forms for consistent UI.
- Redirect all legacy team management routes in both control and orga backends to the unified organizer settings page and update navigation/links accordingly.
- Refine video/world permission mappings, default roles, and trait grants so event-level permissions translate cleanly into video-world capabilities and API behavior.
- Tighten visibility of chats, DMs, admin tools, and room types in the video UI based on the user’s effective permissions.

### CODE REVIEW

File: app/eventyay/webapp/src/components/CreateChatPrompt.vue [hunk 1]
```diff
@@ -53,13 +53,35 @@ export default {
...
 		async create() {
+			// Check if any types are available
+			if (this.types.length === 0) {
+				return
+			}
+
+			// Verify permission for selected type
+			if (this.type === 'text' && !this.hasPermission('world:rooms.create.chat')) {
+				return
+			}
```
[the remark this answers is not in the document]
Comment (Saksham-Sirohi) [reply app/eventyay/webapp/src/components/CreateChatPrompt.vue:84]: correctly implemented

File: app/eventyay/webapp/src/components/CreateDmPrompt.vue [hunk 2]
```diff
@@ -24,6 +24,11 @@ export default {
 	},
 	methods: {
 		async create(users) {
+			// Check permission before creating direct message
+			if (!this.hasPermission('world:chat.direct')) {
```
[the remark this answers is not in the document]
Comment (Saksham-Sirohi) [reply app/eventyay/webapp/src/components/CreateDmPrompt.vue:28]: correctly implemented

**Review by mariobehling — CHANGES_REQUESTED** [review]
@Saksham-Sirohi These are a lot of file changes. The two issues require different changes. So, is it possible to split this PR into two, please? Issue 1 only requires changes to the UI. Issue 2 actually involves some feature changes. So, best would be to keep the PRs separated in the same way the issues are separated. This makes the review clearer as well.

**Comment by hongquan** [comment]
Instead of multiple `if` check like this:

```py
traits_payload = request.data.get("traits") or {}
if not isinstance(traits_payload, dict):
    raise ValidationError("Traits must be provided as an object.")

attendee_trait_grants = traits_payload.get("attendee", "")
if attendee_trait_grants and not isinstance(attendee_trait_grants, str):
    raise ValidationError("Attendee traits must be a string")
```

You should define the expected data shape with Pydantic:

```py
@dataclass
class TraitsPayload:
    attendee: str = 'attendee'

class EventCreationRequestData(BaseModel):
    traits: TraitsPayload
```

then use like this:

```py
from pydantic import ValidationError as PyValidationError
from pyngo import drf_error_details

try:
    req = EventCreationRequestData.model_validate(request.data)
except PyValidationError as e:
    # Convert Pydantic error to DRF error.
    raise drf_error_details(e)
    
req.traits.attendee
```

The benefit of this method is that, you can look into the `EventCreationRequestData` class to know how the data should look like (which fields, data type of each field).

**Comment by norbusan** [comment]
Comments from claude

 Summary & Recommendations

● Overall Assessment: ⚠ NEEDS REVISION BEFORE MERGE

  Verdict: The PR implements valuable features (unified admin, granular video permissions) but has blocking issues that must be addressed.

  ---
  BLOCKING ISSUES 🚨

  Must fix before merge:

  1. Bug in _send_invite: Passes view instance instead of self.request.user to email context
  2. Missing Vue imports: mapGetters imports missing in permission-checking components
  3. TypeError risk: user.traits null safety checks missing
  4. Test coverage: Zero tests for 9 new permission fields and trait mapping
  5. Redirect inconsistency: Team redirect uses section=teams instead of section=permissions

  ---
  HIGH PRIORITY 🔴

  Should address before merge:

  1. View refactoring: Extract team management logic from OrganizerUpdate into mixins/separate views
  2. Permission enforcement: Add backend validation at all video operation endpoints
  3. Audit logging: Verify permission changes are logged
  4. Documentation: Add video permission fields to API docs and developer guide
  5. Migration rollback test: Verify safe rollback to previous schema

  ---
  MEDIUM PRIORITY 🟡

  Address soon after merge:

  1. Performance: Add caching for trait computation if used in hot paths
  2. API versioning: Consider bumping API version or documenting breaking changes
  3. Integration tests: Add end-to-end tests for permission workflows
  4. Frontend tests: Add Vue component tests for permission gating
  5. User docs: Document new permission options for event organizers

  ---
  Recommended Action Plan

  ## Before Re-Review:

  1. Fix blocking bugs:
     - Fix _send_invite user context bug
     - Add missing Vue imports for mapGetters
     - Add null safety for user.traits checks
     - Fix section=teams → section=permissions

  2. Add minimum test coverage:
     - test_video_permissions.py (trait mapping)
     - test_organizer_view.py (team CRUD)
     - test_team_api.py (API serialization)

  3. Security hardening:
     - Add permission checks at video API endpoints
     - Verify audit logging completeness
     - Test migration rollback

  4. Documentation:
     - Update API changelog
     - Add permission field descriptions to docs

  ## After merge (follow-up PR):
  - Refactor OrganizerUpdate into smaller components
  - Add comprehensive integration tests
  - Performance profiling of trait computation

  ---
  Final Notes

  This PR tackles an important consolidation effort and the video permission system is well-designed. However, the implementation has several bugs and gaps that could cause production issues. The Sourcery AI and Copilot feedback aligns with this review - the code needs refinement before merge.

**Comment by norbusan** [comment]
Details from Claude:

● 1. Architecture & Design ✅

  Strengths:
  - Good consolidation approach: Unifying team and organizer management reduces navigation complexity
  - Permission granularity: Nine distinct video permissions provide fine-grained access control
  - Trait-based system: Mapping team permissions to JWT traits is a solid pattern for video integration
  - Backward compatibility: Redirect mixin preserves existing workflows

  Concerns:
  - View bloat: The OrganizerUpdate view is handling too many responsibilities (organizer edit, team CRUD, member management, invites, tokens). This violates Single Responsibility Principle.
  - Missing separation: Consider extracting team operations into a separate TeamManagementView or using Django's ViewSet pattern

  Recommendation:
  # Consider refactoring to:
  class OrganizerUpdate(View):
      # Only organizer-specific logic

  class TeamManagementMixin:
      # Team CRUD operations

  class TeamMemberManagementMixin:
      # Member/invite/token operations

  ---

● 2. Security Analysis ⚠

  Critical Issues:

  1. Potential TypeError with user.traits
    - Code checks user.traits but doesn't handle None case
    - Could cause runtime errors if traits field is null
    - Fix needed: Add null safety checks
  2. Permission escalation risk
    - Staff users automatically receive admin/organizer traits
    - Ensure this is intentional and properly documented
    - Verify that staff status cannot be self-granted
  3. Direct messaging permission bypass
    - Logic conditionally removes DM permissions for organizers
    - Ensure this enforcement happens at API layer, not just UI
    - Need backend validation to prevent API abuse
  4. Audit logging gaps
    - Team operations have audit logging mentioned, but verify completeness
    - Video permission changes should be logged for compliance

  Recommendations:
  - Add explicit permission checks at API endpoints for video operations
  - Implement rate limiting for permission-sensitive operations
  - Add integration tests for permission boundary cases

  ---

● 3. Code Quality 🔧

  Issues Identified:

  1. Sourcery AI feedback (from PR description):
    - _send_invite passes view instance instead of self.request.user - this is a bug
    - Duplicate change-tracking helpers should be consolidated
    - Large view methods need refactoring
  2. Missing imports in Vue components:
    - Copilot identified missing mapGetters imports for hasPermission
    - Will cause runtime errors in production
  3. Inconsistent error handling:
    - Permission error patterns vary across components
    - Should standardize on a common approach
  4. Team redirect logic:
    - Uses section=teams but should use section=permissions based on new UI structure

  Required Fixes:
  # Fix _send_invite to use proper user context
  def _send_invite(self, ...):
      context = {
          'user': self.request.user,  # Not self
          # ...
      }

  ---
  4. Database Migration ✅

  Migration 0003_team_can_video_create_channels_and_more.py:

  Strengths:
  - All new fields have default=False - safe for existing rows
  - Proper verbose names and help text
  - No data transformations - low risk

  Concerns:
  - No rollback test: Ensure migration can be reversed cleanly
  - Index consideration: Nine new boolean fields may benefit from partial indexes if queried frequently
  - Schema size: Adding 9 fields increases table width - acceptable but monitor

  Recommendation:
  - Test migration rollback: python manage.py migrate app_name 0002
  - Consider adding compound index if filtering by multiple video permissions is common:
  indexes = [
      models.Index(fields=['can_video_manage_users', 'can_video_manage_rooms'],
                   name='team_video_mgmt_idx')
  ]

  ---

● 5. API Backward Compatibility ⚠

  Changes to API Serializers:
  - Extended organizer.py serializer with nine new video permission fields
  - Fields are additive only - good for backward compatibility

  Potential Issues:

  1. API version consideration:
    - New fields exposed in /api/v1/ endpoints
    - Existing API clients won't break (fields are optional)
    - But no API version bump - consider documenting in changelog
  2. Permission filtering:
    - Does API properly filter teams based on requesting user's permissions?
    - Verify that users can't read video permissions they shouldn't see
  3. Batch operations:
    - If bulk team updates are supported, ensure video permissions are validated

  Recommendation:
  - Add API tests for new permission fields
  - Document new fields in API schema/OpenAPI spec
  - Verify permission-based filtering in serializer querysets

  ---
  6. Frontend-Backend Integration 🔍

  Key Integration Points:

  1. Permission checks in Vue components:
    - Components gate UI elements behind hasPermission checks
    - Risk: Missing imports could cause undefined behavior
    - Need: Ensure all components importing permission helpers are tested
  2. Trait synchronization:
    - Backend generates JWT traits from team permissions
    - Frontend must decode and respect these traits
    - Verify: Token refresh after permission changes
  3. Real-time updates:
    - Permission changes may not propagate to active sessions immediately
    - Consider: WebSocket notification for permission revocation
  4. Validation mismatch:
    - Frontend filters UI based on permissions
    - Backend MUST enforce same rules at API level
    - Critical: Double-check all video operation endpoints validate permissions

  Recommendation:
  // Ensure consistent permission checks
  // Backend (Django)
  @permission_required('can_video_manage_rooms')
  def create_room(request):
      # ...

  // Frontend (Vue)
  if (hasPermission('can_video_manage_rooms')) {
      // Show create button
  }

  ---

● 7. Testing Requirements ❗

  Critical Gaps:

  Based on the scope of changes (52 files, major permission system overhaul), the PR should include:

  1. Unit tests needed:
    - test_video_permissions.py: Test trait mapping for all 9 permission combinations
    - test_team_model.py: Validate permission field constraints
    - test_organizer_view.py: Test team CRUD operations in unified view
    - test_permission_inheritance.py: Test staff user trait escalation
  2. Integration tests needed:
    - Team creation with video permissions via API
    - Permission enforcement at video endpoints
    - Trait generation and JWT encoding
    - Permission revocation propagation
  3. Frontend tests needed:
    - Component permission gating (Vue Test Utils)
    - Permission helper function coverage
    - UI state updates after permission changes
  4. Migration tests:
    - Forward migration on production-like dataset
    - Rollback migration safety
    - Performance test with large team tables
  5. Regression tests:
    - Legacy URL redirects work correctly
    - Existing team permissions remain functional
    - No permission leakage between organizers

  Recommendation:
  # Minimum test coverage before merge:
  src/tests/base/test_video_permissions.py  # New file
  src/tests/control/test_organizer_unified.py  # New file
  src/tests/api/test_team_api_video_perms.py  # New file

  # Run with:
  py.test src/tests/ -k video_permission -v

  ---

● 8. Additional Concerns

  1. Documentation:
    - No updates to CLAUDE.md or developer docs visible
    - Video permission fields need documenting for future developers
    - API changelog should note new serializer fields
  2. Performance:
    - Nine new boolean fields on Team model - minimal impact
    - Permission caching mentioned but implementation not visible
    - Consider Django's @cached_property for trait computation
  3. User Experience:
    - Redirect from legacy URLs to unified page - good
    - But users may be confused by section parameter inconsistency
    - Need user documentation for new permission granularity

File: app/eventyay/eventyay_common/video/permissions.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    field: str
+    trait_name: str
+
+    def trait_value(self, event_slug: str) -> str:
+        normalized_trait = self.trait_name.replace('_', '-')
+        return f'eventyay-video-event-{event_slug}-{normalized_trait}'
+
+
+VIDEO_PERMISSION_DEFINITIONS: List[VideoPermissionDefinition] = [
```
Comment (hongquan) [inline app/eventyay/eventyay_common/video/permissions.py]: Define video permissions as list is not good, because list allow duplicate, inconsistent entries, like:

```py
[
    VideoPermissionDefinition('can_video_create_stages', 'video_stage_manager'),
    VideoPermissionDefinition('can_video_create_stages', 'video_stages_manager'),
]
```

You should define as `dict` / `'TypedDict`, because `dict` doesn't allow duplicate keys:

```py
{
    'can_video_create_stages': ...
}
```

**Comment by Saksham-Sirohi** [comment]
@sourcery-ai review

**Comment by Saksham-Sirohi** [comment]
https://github.com/user-attachments/assets/3e2e6dd5-7301-40c7-9ba0-1bbb79a33e07

**Comment by hongquan** [comment]
Could you address Copilot comments again?

**Comment by Saksham-Sirohi** [comment]
done

File: app/eventyay/control/navigation.py [hunk 4]
```diff
@@ -475,10 +475,11 @@ def get_organizer_navigation(request):
             {
                 'label': _('Teams'),
                 'url': reverse(
-                    'control:organizer.teams',
+                    'eventyay_common:organizer.update',
                     kwargs={'organizer': request.organizer.slug},
-                ),
-                'active': 'organizer.team' in url.url_name and url.namespace == 'control',
+                )
+                + '?section=permissions',
```
[the remark this answers is not in the document]
Comment (Saksham-Sirohi) [reply app/eventyay/control/navigation.py:481]: false positive

**Comment by Saksham-Sirohi** [comment]
@Sak1012, please test this. This has been long pending

File: app/eventyay/base/migrations/0005_merge_20251229_1317.py
Comment (Sak1012) [inline app/eventyay/base/migrations/0005_merge_20251229_1317.py:1]: Is this migration file necessary it doesn't have any operations.

**Review by Sak1012 — CHANGES_REQUESTED** [review]

File: app/eventyay/base/migrations/0005_team_can_video_create_channels_and_more.py
Comment (Sak1012) [inline app/eventyay/base/migrations/0005_team_can_video_create_channels_and_more.py:1]: `0003` already exists `0003_alter_question_type.py` but this PR is adding another `0003` which is inconsistent. please fix the migrations.

**Comment by Saksham-Sirohi** [comment]
@Sak1012 I have resolved migrations. Please take a look
